handsanitizer/FunctionCallerGenerator.cpp

```diff
@@ -0,0 +1,223 @@
+                output << getJsonInputTemplateTextForJsonRvalue(*arg.getArrayElementType());
+                if (i != arg.getArrayNumElements() - 1) {
+                    output << ",";
+                }
+            }
+            output << "]";
+        }
+
+        if (arg.isStructTy()) {
+            output << "{";
+            for (auto &mem : arg.getNamedMembers()) {
+                output << "\"" << mem.getName() << "\"" << ": "
+                       << getJsonInputTemplateTextForJsonRvalue(*mem.getType());
+                if (mem.getName() != arg.getNamedMembers().back().getName()) { // names should be unique
+                    output << "," << std::endl;
+                }
+            }
+            output << "}";
+        }
+        return output.str();
+    }
+
+    void FunctionCallerGenerator::generate_json_input_template_file(std::string dest_file_path) {
+        std::ofstream of;
+        of.open(dest_file_path, std::ofstream::out | std::ofstream::trunc);
+
+        std::stringstream output;
+        output << "{" << std::endl;
+        output << "\"globals\": {" << std::endl;
+        for (auto &arg: this->declarationManager->globals) {
+            output << "\"" << arg.getName() << "\"" << ": " << getJsonInputTemplateTextForJsonRvalue(*arg.getType());
+            if (arg.getName() != this->declarationManager->globals.back().getName()) {
+                output << "," << std::endl;
+            }
+        }
+
+        output << "}," << std::endl;
+
+        output << "\"arguments\": {" << std::endl;
+        for (auto &arg: function->arguments) {
+            output << "\"" << arg.getName() << "\"" << ": " << getJsonInputTemplateTextForJsonRvalue(*arg.getType());
+            if (arg.getName() != function->arguments.back().getName()) {
+                output << "," << std::endl;
+            }
+        }
+        output << "}" << std::endl;
+
+        output << "}" << std::endl;
+        of << std::setw(4) << output.str() << std::endl;
+        of.close();
+    }
+
+    std::string FunctionCallerGenerator::getMainText() {
+        return R"(
+        int main(int argc, char *argv[]) {
+            setupParser();
+            parser.add_argument("input_file").help("The json file containing arguments");
+            try {
+                parser.parse_args(argc, argv);
+            }
+            catch (const std::runtime_error& err) {
+                std::cout << err.what() << std::endl;
+                std::cout << parser;
+                exit(0);
+            }
+            callFunction();
+            return 0;
+        }
+        )";
+    }
+
+    std::string FunctionCallerGenerator::getFreeVectorFreeText() {
+        std::stringstream output;
+        output << "for(auto& p : " << this->declarationManager->getFreeVectorName() << ") " << std::endl << "free(p);";
+        return output.str();
+    }
+
+    std::vector<GlobalVariable> FunctionCallerGenerator::getGlobals() {
```

Comment:
If callers shouldn't be able to modify the inner object, `const`ify and reference it:

```suggestion
    const std::vector<GlobalVariable>& FunctionCallerGenerator::getGlobals() const {
```